src/HashMap.mo

```diff
@@ -17,12 +17,18 @@ import A "Array";
 import Hash "Hash";
 import Iter "Iter";
 import AssocList "AssocList";
+import Nat32 "Nat32";
 
 module {
 
+  type Key<K> = {
+    // hash field avoids re-hashing the key when the array grows.
+    hash: Hash.Hash;
+    key: K;
+  };
 
   // key-val list type
-  type KVs<K, V> = AssocList.AssocList<K, V>;
+  type KVs<K, V> = AssocList.AssocList<Key<K>, V>;
```

Comment:
Is this type completely hidden from the API? Then I think we can safely change the representation to use a mutable assoc list, with and in-place replace operation. That should cut down on producing lots of garbage when we update buckets of the hash table. Probably best done as a separate PR though.

Reply:
type KVs<K,V> = {key : K; hash : Nat32 ; var value: V ; var tail : KVs<K,V>?} Might perform better. Shame we have to use a record but that's the only way to get  mutable fields.

Reply:
> Is this type completely hidden from the API?

@ggreif asked something related, about upgrades.  Yes, it is.  So it should be possible to do the experiments you sketched with changing it.

> type KVs<K,V> = {key : K; hash : Nat32 ; var value: V ; var tail : KVs<K,V>?} Might perform better. 

> Probably best done as a separate PR though.

Yeah, a more customized bucket list is worth a try, as a separate PR.

In contrast to that future PR, this PR is very targeted: Avoid rehashing.  Fortunately, it turned out to be a small and simple variation over what we had.